pkg/controllers/management/clusterprovisioner/helm_cleaner.go

```diff
@@ -0,0 +1,297 @@
+
+// EmptyHelmDriverName is a placeholder for the empty Helm driver.
+const EmptyHelmDriverName = ""
+
+var (
+	ErrorInvalidMapping = errors.New("invalid API version in mapping")
+)
+
+// generateAPIMappings generates the API mappings for replacement in Helm releases.
+func generateAPIMappings(deprecatedAPIs []DeprecatedAPIData) *mapping.Metadata {
+	var (
+		apiVersionFormat = "apiVersion: %[1]s"
+		kindFormat       = "kind: %[2]s"
+		windowsLineBreak = "\r\n"
+		linuxLineBreak   = "\n"
+
+		formats = []string{
+			apiVersionFormat + linuxLineBreak + kindFormat + linuxLineBreak,     // apiVersion: ... \n kind: ... \n
+			kindFormat + linuxLineBreak + apiVersionFormat + linuxLineBreak,     // kind: ... \n apiVersion: ... \n
+			apiVersionFormat + windowsLineBreak + kindFormat + windowsLineBreak, // apiVersion: ... \r\n kind: ... \r\n
+			kindFormat + windowsLineBreak + apiVersionFormat + windowsLineBreak, // kind: ... \r\n apiVersion: ... \r\n
+		}
+	)
+
+	mappings := mapping.Metadata{}
+
+	for _, api := range deprecatedAPIs {
+		for _, format := range formats {
+			mappingItem := &mapping.Mapping{
+				DeprecatedAPI:    fmt.Sprintf(format, api.DeprecatedAPIVersion, api.Kind),
+				RemovedInVersion: api.KubernetesVersion,
+			}
+
+			if api.NewAPIVersion != "" {
+				mappingItem.NewAPI = fmt.Sprintf(format, api.NewAPIVersion, api.Kind)
+			}
+
+			mappings.Mappings = append(mappings.Mappings, mappingItem)
+		}
+	}
+
+	return &mappings
+}
+
+func newClientGetter(k8sClient kubernetes.Interface, restConfig rest.Config) *wrangler.SimpleRESTClientGetter {
+	return &wrangler.SimpleRESTClientGetter{
+		ClientConfig:    &clientcmd.DefaultClientConfig,
+		RESTConfig:      &restConfig,
+		CachedDiscovery: memory.NewMemCacheClient(k8sClient.Discovery()),
+		RESTMapper:      meta.NewDefaultRESTMapper(nil),
+	}
+}
+
+func (p *Provisioner) cleanupHelmReleases(cluster *v3.Cluster) error {
+	clusterManager := p.clusterManager
+	userContext, err := clusterManager.UserContextNoControllers(cluster.Name)
+	if err != nil {
+		return errors.Wrapf(err, "[cleanupHelmReleases] failed to obtain the Kubernetes client instance")
+	}
+
+	clientGetter := newClientGetter(userContext.K8sClient, userContext.RESTConfig)
+
+	for _, namespace := range FeatureAppNS {
+		actionConfig := &action.Configuration{}
+		if err = actionConfig.Init(clientGetter, namespace, EmptyHelmDriverName, logrus.Infof); err != nil {
+			return errors.Wrapf(err, "[cleanupHelmReleases] failed to create ActionConfiguration instance for Helm")
+		}
+
+		listAction := action.NewList(actionConfig)
+		releases, err := listAction.Run()
+		if err != nil {
+			return errors.Wrapf(err, "[cleanupHelmReleases] failed to list Helm releases for namespace %v", namespace)
+		}
+
+		for _, helmRelease := range releases {
+			lastRelease, err := actionConfig.Releases.Last(helmRelease.Name)
+			if err != nil {
+				logrus.Errorf("[cleanupHelmReleases] failed to find latest release version for release %v", helmRelease.Name)
```

Comment:
If this fails to find the last release it's unlikely anything would change in another run, but I was assuming we would still want to update the releases we can access.
What is the scenario you have in mind?

Reply:
I'm unsure if this ends up contacting the kube-apiserver (maybe it could in the case of a secrets driver), but the scenario that I was envisioning was that the api-server was temporarily unavailable due to an increased request load (i.e. context deadline exceeded or some such).

Reply:
Updated to throw an error.